### PR TITLE
Changes after review - added view model for trams data

### PR DESCRIPTION
**Pull request by mdziubich**

### CODE REVIEW

File: Warsaw_Tram/Extensions/UIViewController.swift [hunk 1]
```diff
@@ -1,37 +1,33 @@
...
 //  Copyright © 2016 Małgorzata Dziubich. All rights reserved.
 //
 
 import UIKit
 
-protocol AlertHelperProtocol {
-    func showAlert(title: String, message: String, okAction: UIAlertAction, cancelAction: UIAlertAction, presentingViewController: UIViewController)
-    func showAlert(title: String, message: String, okAction: UIAlertAction, presentingViewController: UIViewController)
-}
+extension UIViewController {
```
Comment (lsolniczek) [inline Warsaw_Tram/Extensions/UIViewController.swift]: It is not what I mean. Extending object like this is anti-pattern. Go back to protocol `AlertHelperProtocol` and extend it only for `UIViewController` like this:
`extend AlertHelperProtocol where Self: UIViewController`

File: Warsaw_Tram/Extensions/UIViewController.swift [hunk 2]
```diff
@@ -1,37 +1,33 @@
...
         let alertController = UIAlertController(title: title, message: message, preferredStyle: .Alert)
-
+        
         // Add the actions
         alertController.addAction(okAction)
         alertController.addAction(cancelAction)
         
         // Present the controller
-        presentingViewController.presentViewController(alertController, animated: true, completion: nil)
+        self.presentViewController(alertController, animated: true, completion: nil)
```
Comment (lsolniczek) [inline Warsaw_Tram/Extensions/UIViewController.swift]: Don't use `self` when is it not necessary. This is not a mistake, just good practice.

File: Warsaw_Tram/TramViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+//  Warsaw_Tram
+//
+//  Created by Małgorzata Dziubich on 26/06/16.
+//  Copyright © 2016 Małgorzata Dziubich. All rights reserved.
+//
+
+import UIKit
+import CoreLocation
+
+class TramViewModel {
```
Comment (lsolniczek) [inline Warsaw_Tram/TramViewModel.swift]: Good start but this object doesn't looks like ViewModel. I'd rather call it a `Tram`. 
`ViewModel` is a object between `ViewController` and `Model`. `ViewController` is responsible for respond on user action and update view, `Model` responsibility is to serialise data. `ViewModel` is responsible for business logic, what happened when user make a action, how to modify data, and what data should be pass to `ViewController` and present to User.

File: Warsaw_Tram/TramViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        var trams = [Tram]()
+        let stringURL = "https://api.um.warszawa.pl/api/action/wsstore_get?\(resourceId)&\(warsawTramsApiKey)"
+        let url = NSURL(string: stringURL)!
+        
+        let request = NSMutableURLRequest(URL: url)
+        request.HTTPMethod = "GET"
+        
+        let task = NSURLSession.sharedSession().dataTaskWithRequest(request) { data, response, error in
+            
+            if error != nil {
```
Comment (lsolniczek) [inline Warsaw_Tram/TramViewModel.swift]: Use here `guard let` statement to unwrap error if exist and return quick.

File: Warsaw_Tram/TramViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+                failure(error!.localizedDescription)
+            } else if data != nil {
```
Comment (lsolniczek) [inline Warsaw_Tram/TramViewModel.swift]: Use `if let` here

File: Warsaw_Tram/TramViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                guard let jsonResult = (try! NSJSONSerialization.JSONObjectWithData(data!, options: NSJSONReadingOptions.MutableContainers)) as? NSDictionary,
```
Comment (lsolniczek) [inline Warsaw_Tram/TramViewModel.swift]: Write it in right way, don't force `try`, catch error if happend

File: Warsaw_Tram/TramViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+                    let result: NSArray = jsonResult["result"] as? NSArray else {
```
Comment (lsolniczek) [inline Warsaw_Tram/TramViewModel.swift]: Why `NSArray`?

File: Warsaw_Tram/Tram.swift [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+import UIKit
+import CoreLocation
+
+class Tram {
+
+    var number: String!
+    var lattitude: CLLocationDegrees!
+    var longitude: CLLocationDegrees!
+    var lowFloor: Bool!
+    var status: String!
```
Comment (lsolniczek) [inline Warsaw_Tram/Tram.swift]: Why did you use `!` for every class attribute?

File: Warsaw_Tram/TramViewModel.swift [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+//
+//  Created by Małgorzata Dziubich on 28/06/16.
+//  Copyright © 2016 Małgorzata Dziubich. All rights reserved.
+//
+
+import UIKit
+
+protocol TramViewModelProtocol {
+    
+    func getTramsData(success: [Tram] -> Void, failure: String -> Void) -> Void
```
Comment (lsolniczek) [inline Warsaw_Tram/TramViewModel.swift]: You don't need to explicitly return Void value if function doesn't return value

File: Warsaw_Tram/Tram.swift (hunk 8 above)
Comment (mdziubich) [inline Warsaw_Tram/Tram.swift]: To note that this values can't be nil. That's how I've been taught.

Comment (lsolniczek) [inline Warsaw_Tram/Tram.swift]: This is wrong. To note that a parameter can not be `nil` type is just enough. When you use `!` you are saying to the compiler that I don't what to set parameter during object initialisation but you are sure that when you use them first time there will be some value (not nil). You are setting this values in init, I'd rather remove `!`. You can read more about `implicitly unwrapped` and `optional paramter`

File: Warsaw_Tram/TramViewModel.swift (hunk 7 above)
Comment (mdziubich) [inline Warsaw_Tram/TramViewModel.swift]: I just got an error when using Array, but now I changed it and it looks like this:
`let result: Array = jsonResult["result"] as? Array<AnyObject>`

File: Warsaw_Tram/MapViewController.swift [hunk 10]
```diff
@@ -47,46 +46,31 @@ class MapViewController: UIViewController {
...
-                            let lowFloor = tram["LowFloor"] as? Bool {
-                                dispatch_async(dispatch_get_main_queue(), { 
-                                    self.addTramAnnotation(line, lat: lat, lon: lon, lowFloor: lowFloor)
-                                })
-                        }
-                    }
+    private func displayTramsOnMap() {
+        tramViewModel.getTramsData({ (trams) in
+            for tram in trams {
+                if tram.status == "RUNNING" {
```
Comment (lsolniczek) [inline Warsaw_Tram/MapViewController.swift]: Instead of another level of nesting you can use `for tram in trams where tram.status == "RUNNING"`

**Comment by lsolniczek** [comment]
Looks good. You can merge it 👍

**Pull request closed without merging**